### PR TITLE
feat: add support for CycloneDX 1.4 specification

### LINKED ISSUE #293: [FEATURE] Support for version 1.4 schema
CycloneDX schema version 1.4 is due for release in the coming weeks.

This feature will add support to generate SBOM's using `cyclonedx-python` in the latest 1.4 format.

### PR DESCRIPTION
**Pull request by madpah**
feat: add support for output to CycloneDX 1.4 (draft)
feat: Error with return code 2 if attempting to output in JSON and SchemaVersion < 1.2
test: Multiple tests added

Closes #293, #280 

Signed-off-by: Paul Horton <phorton@sonatype.com>

### CODE REVIEW

File: tests/base.py [hunk 1]
```diff
@@ -16,15 +16,25 @@
...
 from datetime import datetime, timezone
+from lxml import etree
+from os import path
+from typing import Any
 from unittest import TestCase
 from uuid import uuid4
 from xml.dom import minidom
-from typing import Any
+
+from lxml.etree import DocumentInvalid
```
[the remark this answers is not in the document]
Comment (madpah) [reply tests/base.py:32]: @sonatype-lift ignore

File: tests/base.py [hunk 2]
```diff
@@ -81,6 +114,29 @@ def assertEqualJsonBom(self, a: str, b: str) -> None:
...
+
+    def assertValidAgainstSchema(self, bom_xml: str, schema_version: SchemaVersion) -> None:
+        xsd_fn = os.path.join(schema_directory, f'bom-{schema_version.name.replace("_", ".").replace("V", "")}.xsd')
+        with open(xsd_fn) as xsd_fd:
+            xsd_doc = etree.parse(xsd_fd)
+
+        xml_schema = etree.XMLSchema(xsd_doc)
+        schema_validates = False
+        try:
+            schema_validates = xml_schema.validate(etree.parse(io.BytesIO(bytes(bom_xml, 'ascii'))))
```
[the remark this answers is not in the document]
Comment (madpah) [reply tests/base.py:129]: @sonatype-lift ignore

File: tests/base.py [hunk 3]
```diff
@@ -16,15 +16,25 @@
...
 # SPDX-License-Identifier: Apache-2.0
 # Copyright (c) OWASP Foundation. All Rights Reserved.
-
+import io
 import json
+import os
 import sys
 import xml.etree.ElementTree
 from datetime import datetime, timezone
+from lxml import etree
```
[the remark this answers is not in the document]
Comment (madpah) [reply tests/base.py:25]: @sonatype-lift ignore

File: tests/test_cyclonedx_json.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) OWASP Foundation. All Rights Reserved.
+
+from os import path
+import subprocess
```
[the remark this answers is not in the document]
Comment (madpah) [reply tests/test_cyclonedx_json.py:21]: @sonatype-lift ignore

File: cyclonedx_py/client.py [hunk 5]
```diff
@@ -100,6 +101,14 @@ def get_output(self) -> BaseOutput:
         )
 
     def execute(self) -> None:
+        # Quick check for JSON && SchemaVersion <= 1.1
+        if str(self._arguments.output_format).upper() == 'JSON' and \
```
Comment (jkowalleck) [inline cyclonedx_py/client.py:105]: :heart_decoration:

**Review by jkowalleck — CHANGES_REQUESTED** [review]
i dont see why adding all the schema files to the project and doinng tests on JSON/XML is part of this project.

**Comment by jkowalleck** [comment]
@madpah  re https://github.com/CycloneDX/cyclonedx-python/pull/294#pullrequestreview-847148710
why add the JSON/XML/proto schema to the project? why ship it in the package?  

Dont see a reason why the files should be part of this package's tests, as they are part of [cyclonedx-python-lib](https://github.com/CycloneDX/cyclonedx-python-lib/tree/feat/schema-version-1.4/cyclonedx/schema) already. 
you could access the resources inside a package - via [importlib](https://docs.python.org/3.8/library/importlib.html?highlight=importlib#module-importlib.resources) or [pkg](https://setuptools.pypa.io/en/latest/pkg_resources.html#basic-resource-access)

File: tests/base.py [hunk 6]
```diff
@@ -36,10 +46,33 @@
...
+                schema_doc = json.load(schema_fd)
+
+            try:
+                json_validate(instance=json.loads(bom_json), schema=schema_doc)
+            except ValidationError as e:
+                self.assertTrue(False, f'Failed to validate SBOM against JSON schema: {str(e)}')
+
+            self.assertTrue(True)
+        else:
+            self.assertTrue(True, 'JSON Schema Validation is not possible in Python < 3.7')
```
Comment (jkowalleck) [inline tests/base.py:74]: use unittest's `skipif` https://docs.python.org/3/library/unittest.html?highlight=skipif#unittest.skipIf
to skip tests. sont write conditional foo

File: tests/base.py [hunk 7]
```diff
@@ -81,6 +114,29 @@ def assertEqualJsonBom(self, a: str, b: str) -> None:
...
+            schema_validates = xml_schema.validate(etree.parse(io.BytesIO(bytes(bom_xml, 'ascii'))))
+        except DocumentInvalid as e:
+            print(f'Failed to validate SBOM against schema: {str(e)}')
+        except Exception as e:
+            print(f'Exception: {str(e)}')
+            print(f'BOM XML is: {bom_xml}')
+
+        if not schema_validates:
+            print(xml_schema.error_log.last_error)
+        self.assertTrue(schema_validates, 'Failed to validate Generated SBOM against XSD Schema')
```
Comment (jkowalleck) [inline tests/base.py:138]:
```patch
-        if not schema_validates:
-            print(xml_schema.error_log.last_error)
-        self.assertTrue(schema_validates, 'Failed to validate Generated SBOM against XSD Schema')
+        self.assertTrue(schema_validates, 'Failed to validate Generated SBOM against XSD Schema\n'+xml_schema.error_log.last_error)
```

File: tests/base.py [hunk 8]
```diff
@@ -81,6 +114,29 @@ def assertEqualJsonBom(self, a: str, b: str) -> None:
...
+    def assertValidAgainstSchema(self, bom_xml: str, schema_version: SchemaVersion) -> None:
+        xsd_fn = os.path.join(schema_directory, f'bom-{schema_version.name.replace("_", ".").replace("V", "")}.xsd')
+        with open(xsd_fn) as xsd_fd:
+            xsd_doc = etree.parse(xsd_fd)
+
+        xml_schema = etree.XMLSchema(xsd_doc)
+        schema_validates = False
+        try:
+            schema_validates = xml_schema.validate(etree.parse(io.BytesIO(bytes(bom_xml, 'ascii'))))
+        except DocumentInvalid as e:
```
Comment (jkowalleck) [inline tests/base.py:130]: whyyyyy? 
why catch the exception, if this exception is clearly a reason that this test would fail?

**Review by jkowalleck — CHANGES_REQUESTED** [review]

**Comment by madpah** [comment]
@jkowalleck - fair point on the schema validation comments - will look at moving these (where required) back to `cyclonedx-python-lib`.

**Comment by madpah** [comment]
> @madpah re [#294 (review)](https://github.com/CycloneDX/cyclonedx-python/pull/294#pullrequestreview-847148710) why add the JSON/XML/proto schema to the project? why ship it in the package?
> 
> Dont see a reason why the files should be part of this package's tests, as they are part of [`cyclonedx-python-lib`](https://github.com/CycloneDX/cyclonedx-python-lib/tree/feat/schema-version-1.4/cyclonedx/schema) already. you could access the resources inside a package - via [`importlib`](https://docs.python.org/3.8/library/importlib.html?highlight=importlib#module-importlib.resources) or [`pkg`](https://setuptools.pypa.io/en/latest/pkg_resources.html#basic-resource-access)
> 
> Saw the files are used in tests, to check output of `cyclonedx-python-lib` so why not add a `validate` capability to `cyclonedx-python-lib` ? it seams to be the perfect place.
> 
> ❌ what were the cons? why did you decide to put the data-validation in a downstream CLI tool instead of the library that generates the data in the first place?

Totally agree - have removed as covered in `cylonedx-python-lib`.

**Pull request merged**